Sources/Workspace/SourceControlPackageContainer.swift

```diff
@@ -142,6 +149,68 @@ internal final class SourceControlPackageContainer: PackageContainer, CustomStri
         return try? self.knownVersions()[version]
     }
 
+    func checkIntegrity(version: Version, revision: Revision) throws {
+        guard case .remoteSourceControl(let sourceControlURL) = self.package.kind else {
+            return
+        }
+        
+        let packageIdentity = self.package.identity
+        let fingerprint: Fingerprint
+        do {
+            fingerprint = try temp_await {
+                self.fingerprintStorage.get(
+                    package: packageIdentity,
+                    version: version,
+                    kind: .sourceControl,
+                    observabilityScope: self.observabilityScope,
+                    callbackQueue: .sharedConcurrent,
+                    callback: $0
+                )
+            }
+        } catch PackageFingerprintStorageError.notFound {
+            fingerprint = Fingerprint(origin: .sourceControl(sourceControlURL), value: revision.identifier)
+            // Write to storage if fingerprint not yet recorded
+            do {
+                try temp_await {
+                    self.fingerprintStorage.put(
+                        package: packageIdentity,
+                        version: version,
+                        fingerprint: fingerprint,
+                        observabilityScope: self.observabilityScope,
+                        callbackQueue: .sharedConcurrent,
+                        callback: $0
+                    )
+                }
+            } catch PackageFingerprintStorageError.conflict(_, let existing) {
+                let message = "Revision \(revision.identifier) for \(self.package) version \(version) does not match previously recorded value \(existing.value) from \(String(describing: existing.origin.url?.absoluteString))"                
+                switch self.fingerprintCheckingMode {
+                case .strict:
+                    throw StringError(message)
+                case .warn:
+                    observabilityScope.emit(warning: message)
+                case .none:
+                    return
+                }
+            }
+        } catch {
+            self.observabilityScope.emit(error: "Failed to get source control fingerprint for \(self.package) version \(version) from storage: \(error)")
+            throw error
+        }
+        
+        // The revision (i.e., hash) must match that in fingerprint storage otherwise the integrity check fails
+        if revision.identifier != fingerprint.value {
```

Comment:
should this check move to L169 (inside the first case basically)?

Reply:
It could be, but then any error thrown would get caught in the `catch` block (if you recall the code was checking for `StringError` specifically) and the code wouldn't be as clean.